drivers/hmis/app/models/hmis/user.rb

```diff
@@ -115,13 +115,21 @@ def permissions_for?(entity, *permissions, mode: :any)
     check_permissions_with_mode(*permissions, mode: mode) { |perm| permission_for?(entity, perm) }
   end
 
-  def entities_with_permissions(model, *permissions, **kwargs)
-    model.where(
-      id: Hmis::GroupViewableEntity.where(
-        access_group_id: access_groups.with_permissions(*permissions, **kwargs).pluck(:id),
-        entity_type: model.sti_name,
-      ).select(:entity_id),
-    )
+  def entities_with_permissions(model, *permissions, mode: :any)
+    # Get all the roles that have this permission
+    roles_with_permission = Hmis::Role.with_permissions(*permissions, mode: mode).pluck(:id)
+
+    # Find the Access Controls that this user is assigned to that have any of the approved Roles,
+    # and pluck the Access Group (aka Collection). The user has access <permission>-level access for
+    # any entity in these access groups.
+    access_group_ids = access_controls.where(role_id: roles_with_permission).pluck(:access_group_id)
```

Comment:
Replacing this line with the following breaks the tests
```
    access_group_ids = access_groups.joins(access_controls: :role).merge(Hmis::Role.with_permissions(*permissions, mode: mode)).pluck(:id)
```

could be something to do with relationships between models not being right..